Node v14+ support and fix for TypeScript and EventEmitter

Node v14 is the active LTS and it switches to "maintenance" tomorrow, making Node v16 the current LTS as of Tuesday. It would be great to be able to upgrade the project so it builds without errors. When I try, I get the following:

```
node_modules/@fluent-org/logger/build/src/server.d.ts:63:43 - error TS2507: Type 'typeof EventEmitter' is not a constructor function type.

63 export declare class FluentServer extends EventEmitter {
                                             ~~~~~~~~~~~~

node_modules/@fluent-org/logger/build/src/socket.d.ts:164:43 - error TS2507: Type 'typeof EventEmitter' is not a constructor function type.

164 export declare class FluentSocket extends EventEmitter {
                                              ~~~~~~~~~~~~


Found 2 errors.
```

From what I see at [DefinitelyTyped](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/41353#issuecomment-579381008), the upgrade at 13.4 or 13.5 caused issues. [Geckos.io](https://github.com/geckosio/geckos.io/issues/22) has a simple fix, as does [cheap-watch](https://githubmemory.com/repo/Conduitry/cheap-watch/issues/4). It basically means you have to change the import line.

```
// Old
import * as EventEmitter from "events";

// New
import { EventEmitter } from 'events';
```

--- a/package.json
+++ b/package.json
@@ -30,24 +30,23 @@
     "node": ">=12.9.0"
   },
   "dependencies": {
-    "@msgpack/msgpack": "^2.7.0",
-    "p-defer": "^3.0.0"
+    "@msgpack/msgpack": "^2.7.1"
   },
   "devDependencies": {
-    "@types/chai": "^4.2.19",
+    "@types/chai": "^4.2.22",
     "@types/chai-as-promised": "^7.1.4",
-    "@types/duplexer3": "^0.1.0",
-    "@types/mocha": "^8.2.2",
-    "@types/node": "^14.11.2",
-    "@types/sinon": "^10.0.2",
+    "@types/duplexer3": "^0.1.1",
+    "@types/mocha": "^9.0.0",
+    "@types/node": "^16.11.10",
+    "@types/sinon": "^10.0.6",
     "chai": "^4.3.4",
     "chai-as-promised": "^7.1.1",
     "duplexer3": "^0.1.4",
     "eslint": "^7.28.0",
     "gts": "^3.1.0",
-    "mocha": "^9.0.1",
-    "sinon": "^11.1.1",
-    "typedoc": "^0.21.0",
+    "mocha": "^9.1.3",
+    "sinon": "^12.0.1",
+    "typedoc": "^0.22.9",
     "typescript": "^4.0.3"
   },
   "license": "Apache-2.0",

--- a/src/auth.ts
+++ b/src/auth.ts
@@ -161,7 +161,7 @@ export class FluentAuthSocket extends FluentSocket {
       this.authState = FluentAuthState.AUTHENTICATED;
       this.onEstablished();
     } catch (e) {
-      return this.close(CloseState.FATAL, e);
+      return this.close(CloseState.FATAL, e as Error);
     }
   }
 }

--- a/src/client.ts
+++ b/src/client.ts
@@ -16,10 +16,14 @@ import {
   MessageQueue,
   ForwardQueue,
 } from "./modes";
-import {DeferredPromise} from "p-defer";
 import * as crypto from "crypto";
 import {EventRetrier, EventRetryOptions} from "./event_retrier";
-import {awaitAtMost, awaitNextTick, awaitTimeout} from "./util";
+import {
+  DeferredPromise,
+  awaitAtMost,
+  awaitNextTick,
+  awaitTimeout,
+} from "./util";
 
 type AckData = {
   timeoutId: NodeJS.Timeout;

--- a/src/event_retrier.ts
+++ b/src/event_retrier.ts
@@ -1,6 +1,5 @@
-import * as pDefer from "p-defer";
 import {DroppedError, RetryShutdownError} from "./error";
-import {awaitAtMost} from "./util";
+import {awaitAtMost, pDefer, DeferredPromise} from "./util";
 
 /**
  * Event retry settings
@@ -48,7 +47,7 @@ export type EventRetryOptions = {
  */
 export class EventRetrier {
   private options: EventRetryOptions;
-  private cancelWait: pDefer.DeferredPromise<void>;
+  private cancelWait: DeferredPromise<void>;
   constructor(opts: Partial<EventRetryOptions> = {}) {
     this.options = {
       attempts: 4,
@@ -120,7 +119,7 @@ export class EventRetrier {
           throw e;
         }
 
-        this.options.onError(e);
+        this.options.onError(e as Error);
 
         const retryInterval = Math.min(
           this.options.maxDelay,

--- a/src/modes/forward.ts
+++ b/src/modes/forward.ts
@@ -1,11 +1,11 @@
-import * as pDefer from "p-defer";
 import {Queue, PacketData} from "./queue";
 import * as protocol from "../protocol";
+import {pDefer, DeferredPromise} from "../util";
 
 type ForwardRecord = {
   tag: protocol.Tag;
   entries: protocol.Entry[];
-  deferred: pDefer.DeferredPromise<void>;
+  deferred: DeferredPromise<void>;
 };
 
 /**

--- a/src/modes/message.ts
+++ b/src/modes/message.ts
@@ -1,12 +1,12 @@
-import * as pDefer from "p-defer";
 import {Queue, PacketData} from "./queue";
 import * as protocol from "../protocol";
+import {pDefer, DeferredPromise} from "../util";
 
 type EventRecord = {
   tag: protocol.Tag;
   time: protocol.Time;
   event: protocol.EventRecord;
-  deferred: pDefer.DeferredPromise<void>;
+  deferred: DeferredPromise<void>;
 };
 
 /**

--- a/src/modes/packed_forward.ts
+++ b/src/modes/packed_forward.ts
@@ -1,6 +1,6 @@
-import * as pDefer from "p-defer";
 import {PacketData, Queue} from "./queue";
 import * as protocol from "../protocol";
+import {pDefer, DeferredPromise} from "../util";
 
 type PackedRecord = {
   tag: protocol.Tag;
@@ -11,7 +11,7 @@ type PackedRecord = {
    * Useful for concatenating them all together later
    */
   size: number;
-  deferred: pDefer.DeferredPromise<void>;
+  deferred: DeferredPromise<void>;
 };
 
 /**

--- a/src/modes/queue.ts
+++ b/src/modes/queue.ts
@@ -1,6 +1,6 @@
 import * as protocol from "../protocol";
-import {DeferredPromise} from "p-defer";
 import {DroppedError, QueueShutdownError} from "../error";
+import {DeferredPromise} from "../util";
 
 /**
  * Every queue must have this type of data

--- a/src/server.ts
+++ b/src/server.ts
@@ -1,7 +1,7 @@
 import * as net from "net";
 import * as tls from "tls";
 import * as crypto from "crypto";
-import * as EventEmitter from "events";
+import {EventEmitter} from "events";
 import * as protocol from "./protocol";
 import {UnexpectedMessageError} from "./error";
 
@@ -190,7 +190,7 @@ export class FluentServer extends EventEmitter {
             const pong = protocol.generatePong(
               this.security.serverHostname,
               false,
-              e.message,
+              (e as Error).message,
               clientInfo.sharedKeyInfo
             );
             socket.write(protocol.encodeMessage(pong));

--- a/src/socket.ts
+++ b/src/socket.ts
@@ -1,6 +1,6 @@
 import * as net from "net";
 import * as tls from "tls";
-import * as EventEmitter from "events";
+import {EventEmitter} from "events";
 import {
   SocketTimeoutError,
   UnexpectedMessageError,

--- a/src/util.ts
+++ b/src/util.ts
@@ -24,3 +24,18 @@ export const awaitNextTick = (): Promise<void> => {
 export const awaitTimeout = (timeout: number): Promise<void> => {
   return new Promise(r => setTimeout(r, timeout));
 };
+
+export interface DeferredPromise<T> {
+  promise: Promise<T>;
+  resolve(value?: T | PromiseLike<T>): void;
+  reject(reason?: unknown): void;
+}
+
+export const pDefer = <T>(): DeferredPromise<T> => {
+  const deferred: Partial<DeferredPromise<T>> = {};
+  deferred.promise = new Promise((resolve, reject) => {
+    deferred.resolve = resolve;
+    deferred.reject = reject;
+  });
+  return deferred as DeferredPromise<T>;
+};

--- a/test/helpers.ts
+++ b/test/helpers.ts
@@ -1,4 +1,4 @@
-import * as EventEmitter from "events";
+import {EventEmitter} from "events";
 import {Duplex, PassThrough, Readable, Writable} from "stream";
 import * as duplexer3 from "duplexer3";
 

--- a/test/test.event_retrier.ts
+++ b/test/test.event_retrier.ts
@@ -1,9 +1,9 @@
 import * as chai from "chai";
 import * as chaiAsPromised from "chai-as-promised";
 import * as sinon from "sinon";
-import * as pDefer from "p-defer";
 import {EventRetrier} from "../src/event_retrier";
 import {RetryShutdownError} from "../src/error";
+import {pDefer} from "../src/util";
 
 chai.use(chaiAsPromised);
 const expect = chai.expect;

